File: frontend/src/ink.js
// Handwriting: the client half of gamma/ink.py (docs/dev/handwriting.md).
// Pure — no React, no network — so it runs under node --test.
//
// An ink group is one `gamma-ink` file: strokes on one PDF page, in the
// page's scale-1 frame (pdf.js viewport at scale 1: points, origin
// top-left, rotation applied — the same frame highlight rects normalise
// to). Samples are stored per stroke as one flat integer array, `ch`
// naming the channels ("xy" + any of p/t/a/z); x, y (1/100 pt) and t (ms)
// are deltas after the first sample, p is 0..1000. Keep the codec here in
// step with the backend's.
import { getStroke } from "perfect-freehand";

export const FORMAT = "gamma-ink";
export const VERSION = 1;
export const COORD_UNIT = 100;
export const PRESSURE_UNIT = 1000;
export const THINNING = 0.5;      // width = size * (1 + THINNING * (p - 0.5)); mirrors ink.py
export const MAX_STROKES = 5000;

// Pen sizes (pt at scale 1, S/M/L) and highlighter sizes.
export const PEN_SIZES = [1.2, 2, 3.2];
export const HIGHLIGHTER_SIZES = [8, 12, 18];
export const PEN_COLORS = ["#1f1f1f", "#1d4ed8", "#dc2626", "#15803d", "#7c3aed", "#ea580c"];

export function newInk(page, width, height) {
  return { format: FORMAT, version: VERSION,
    space: { kind: "pdf-page", page, width, height }, strokes: [] };
}

const ID_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
export function strokeId() {
  let s = "";
  for (let i = 0; i < 8; i++) s += ID_CHARS[Math.floor(Math.random() * ID_CHARS.length)];
  return s;
}

// --- codec -----------------------------------------------------------------

// samples: [{x, y, p?, t?}] with x/y in points and t in ms since t0.
export function encodeStroke({ id, tool = "pen", color = PEN_COLORS[0], size = 2, opacity = 1,
  pen = true, t0 = null, samples, ch = "xyp" }) {
  const pts = [];
  let px = 0, py = 0, pt = 0;
  for (const s of samples) {
    for (const c of ch) {
      if (c === "x") { const v = Math.round(s.x * COORD_UNIT); pts.push(v - px); px = v; }
      else if (c === "y") { const v = Math.round(s.y * COORD_UNIT); pts.push(v - py); py = v; }
      else if (c === "p") pts.push(Math.round(Math.max(0, Math.min(1, s.p ?? 0.5)) * PRESSURE_UNIT));
      else if (c === "t") { const v = Math.round(s.t || 0); pts.push(v - pt); pt = v; }
      else if (c === "a") pts.push(Math.round(s.a || 0));
      else if (c === "z") pts.push(Math.round(s.z || 0));
    }
  }
  const out = { id: id || strokeId(), tool, color, size, opacity, pen, ch, pts };
  if (t0 != null) out.t0 = t0;
  return out;
}

const decoded = new WeakMap();
// → [{x, y, p, t}] in points; p defaults to 0.5, t to null. Cached per stroke object.
export function decodeStroke(stroke) {
  const hit = decoded.get(stroke);
  if (hit) return hit;
  const { ch = "xy", pts = [] } = stroke;
  const n = ch.length, out = [];
  let x = 0, y = 0, t = 0;
  for (let i = 0; i + n <= pts.length; i += n) {
    const s = { x: 0, y: 0, p: 0.5, t: null };
    for (let k = 0; k < n; k++) {
      const c = ch[k], v = pts[i + k];
      if (c === "x") { x += v; s.x = x / COORD_UNIT; }
      else if (c === "y") { y += v; s.y = y / COORD_UNIT; }
      else if (c === "p") s.p = Math.max(0, Math.min(1, v / PRESSURE_UNIT));
      else if (c === "t") { t += v; s.t = t; }
      else if (c === "a") s.a = v;
      else if (c === "z") s.z = v;
    }
    out.push(s);
  }
  decoded.set(stroke, out);
  return out;
}

export function strokeWidth(stroke, p) {
  if (stroke.tool !== "pen" || stroke.pen === false) return stroke.size;
  return stroke.size * (1 + THINNING * (p - 0.5));
}

// --- geometry --------------------------------------------------------------

const bounds = new WeakMap();
// [x0, y0, x1, y1] around the samples, width included.
export function strokeBounds(stroke) {
  const hit = bounds.get(stroke);
  if (hit) return hit;
  let x0 = Infinity, y0 = Infinity, x1 = -Infinity, y1 = -Infinity;
  for (const s of decodeStroke(stroke)) {
    const r = strokeWidth(stroke, s.p) / 2;
    x0 = Math.min(x0, s.x - r); y0 = Math.min(y0, s.y - r);
    x1 = Math.max(x1, s.x + r); y1 = Math.max(y1, s.y + r);
  }
  const out = x0 === Infinity ? null : [x0, y0, x1, y1];
  bounds.set(stroke, out);
  return out;
}

export function inkBounds(ink) {
  let out = null;
  for (const s of ink?.strokes || []) {
    const b = strokeBounds(s);
    if (!b) continue;
    out = out ? [Math.min(out[0], b[0]), Math.min(out[1], b[1]), Math.max(out[2], b[2]), Math.max(out[3], b[3])] : [...b];
  }
  return out;
}

// The group's box in the highlight `pdf_position` shape (what jump-to-
// position, the sidebar marker and the exporters read).
export function pdfPositionOf(ink) {
  const b = inkBounds(ink);
  if (!b || ink.space?.kind !== "pdf-page") return null;
  const r2 = (v) => Math.round(v * 100) / 100;
  const rect = { x1: r2(b[0]), y1: r2(b[1]), x2: r2(b[2]), y2: r2(b[3]),
    width: ink.space.width, height: ink.space.height, pageNumber: ink.space.page };
  return { pageNumber: ink.space.page, boundingRect: rect, rects: [{ ...rect }] };
}

function segDist2(px, py, ax, ay, bx, by) {
  const dx = bx - ax, dy = by - ay;
  const len2 = dx * dx + dy * dy;
  let t = len2 ? ((px - ax) * dx + (py - ay) * dy) / len2 : 0;
  t = Math.max(0, Math.min(1, t));
  const qx = ax + t * dx - px, qy = ay + t * dy - py;
  return qx * qx + qy * qy;
}

// Ids of the strokes an eraser at (x, y) with `radius` touches: within
// half the stroke's width plus the radius of any segment. Bounding boxes
// prune the work.
export function hitStrokes(ink, x, y, radius) {
  const out = [];
  for (const s of ink?.strokes || []) {
    const b = strokeBounds(s);
    if (!b || x < b[0] - radius || x > b[2] + radius || y < b[1] - radius || y > b[3] + radius) continue;
    const pts = decodeStroke(s);
    let hit = false;
    for (let i = 0; i < pts.length && !hit; i++) {
      const a = pts[i], c = pts[Math.min(i + 1, pts.length - 1)];
      const tol = radius + Math.max(strokeWidth(s, a.p), strokeWidth(s, c.p)) / 2;
      if (segDist2(x, y, a.x, a.y, c.x, c.y) <= tol * tol) hit = true;
    }
    if (hit) out.push(s.id);
  }
  return out;
}

export function appendStroke(ink, stroke) {
  return { ...ink, strokes: [...(ink.strokes || []), stroke].slice(-MAX_STROKES) };
}

export function removeStrokes(ink, ids) {
  const drop = new Set(ids);
  return { ...ink, strokes: (ink.strokes || []).filter((s) => !drop.has(s.id)) };
}

// --- rendering -------------------------------------------------------------

const avg = (a, b) => (a + b) / 2;
// perfect-freehand's README recipe: the outline polygon as a closed path of
// quadratic curves through midpoints (smooth, no visible corners).
export function svgPathFromPoints(points) {
  const n = points.length;
  if (!n) return "";
  if (n < 3) {
    const [a] = points;
    return `M${a[0].toFixed(2)},${a[1].toFixed(2)} L${(a[0] + 0.01).toFixed(2)},${a[1].toFixed(2)}`;
  }
  let d = `M${points[0][0].toFixed(2)},${points[0][1].toFixed(2)} Q`;
  for (let i = 0; i < n; i++) {
    const a = points[i], b = points[(i + 1) % n];
    d += `${a[0].toFixed(2)},${a[1].toFixed(2)} ${avg(a[0], b[0]).toFixed(2)},${avg(a[1], b[1]).toFixed(2)} `;
  }
  return d + "Z";
}

// Outline options in page units — thickness is size, so zoom scales ink
// like ink on paper.
export function outlineOptions(stroke) {
  return { size: stroke.size, thinning: stroke.pen === false ? 0 : THINNING, smoothing: 0.5,
    streamline: 0.4, simulatePressure: false, last: true };
}

// samples → the filled outline path for a pen stroke.
export function penOutline(samples, opts) {
  return svgPathFromPoints(getStroke(samples.map((s) => [s.x, s.y, s.p ?? 0.5]), opts));
}

const paths = new WeakMap();
// The SVG rendering of a stroke: {d, fill} (pen: a filled outline) or
// {d, stroke, width} (highlighter: a stroked polyline). Cached per object.
export function strokePath(stroke) {
  const hit = paths.get(stroke);
  if (hit) return hit;
  const pts = decodeStroke(stroke);
  let out;
  if (stroke.tool === "highlighter") {
    const d = pts.map((s, i) => `${i ? "L" : "M"}${s.x.toFixed(2)},${s.y.toFixed(2)}`).join(" ")
      + (pts.length === 1 ? ` L${(pts[0].x + 0.01).toFixed(2)},${pts[0].y.toFixed(2)}` : "");
    out = { d, stroke: true, width: stroke.size };
  } else {
    out = { d: penOutline(pts, outlineOptions(stroke)), stroke: false };
  }
  paths.set(stroke, out);
  return out;
}

File: backend/gamma/ink.py
"""Handwriting: the ``gamma-ink`` stroke file and the pure geometry every
consumer shares (docs/dev/handwriting.md).

An ink group is a block whose ``properties.ink_url`` names one of these
files (``uploads/<sha>.ink``, plain JSON). The file is the only source of
truth for the strokes; nothing derived is stored. Coordinates are the page
as displayed at scale 1 (pdf.js viewport: points, origin top-left, y down,
rotation applied) — the same frame highlight rects normalise to, so the PDF
writers map ink through the exact conversion they already use for rects.

    {"format": "gamma-ink", "version": 1,
     "space": {"kind": "pdf-page", "page": 3, "width": 612, "height": 792},
     "strokes": [{"id": "k7Qm2x", "tool": "pen", "color": "#1f1f1f",
                  "size": 1.6, "opacity": 1, "pen": true, "t0": 1757760000000,
                  "ch": "xypt", "pts": [12040, 30512, 620, 0, 18, -3, 700, 8]}]}

``ch`` declares the channels of every sample (InkML's idea): ``x`` ``y``
always, then any of ``p`` pressure, ``t`` time, ``a`` altitude, ``z``
azimuth. ``pts`` is one flat integer array — x/y in 1/100 pt and t in ms
are delta-encoded after the first sample, p is 0..1000, a/z degrees. The
frontend's ``src/ink/ink.js`` is the mirror of the codec here; keep them in
step.
"""

import json
import math
import re
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError, model_validator

FORMAT = "gamma-ink"
VERSION = 1
MAX_STROKES = 5000
MAX_SAMPLES = 500_000
MAX_BYTES = 4 * 1024 * 1024
COORD_UNIT = 100        # stored x/y are hundredths of a point
PRESSURE_UNIT = 1000    # stored p is 0..1000
THINNING = 0.5          # width = size * (1 + THINNING * (p - 0.5)); mirrors ink.js
_CH_RE = re.compile(r"^xy(?!.*(.).*\1)[ptaz]*$")
_COLOR_RE = re.compile(r"^(#[0-9a-fA-F]{6}|rgba?\(\s*\d+\s*,\s*\d+\s*,\s*\d+\s*(,\s*[0-9.]+\s*)?\))$")


class Space(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)
    kind: Literal["pdf-page", "canvas"] = "pdf-page"
    page: int | None = Field(default=None, ge=1)
    width: float = Field(gt=0, le=100_000)
    height: float = Field(gt=0, le=100_000)

    @model_validator(mode="after")
    def _page_for_pdf(self):
        if self.kind == "pdf-page" and self.page is None:
            raise ValueError("a pdf-page space needs a page number")
        return self


class Stroke(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)
    id: str = Field(min_length=1, max_length=32, pattern=r"^[A-Za-z0-9_-]+$")
    tool: Literal["pen", "highlighter"] = "pen"
    color: str = Field(default="#1f1f1f", max_length=40)
    size: float = Field(default=1.6, gt=0, le=100)
    opacity: float = Field(default=1.0, gt=0, le=1)
    pen: bool = True
    t0: int | None = Field(default=None, ge=0)
    ch: str = Field(default="xy", max_length=6)
    pts: list[int] = Field(default_factory=list)

    @model_validator(mode="after")
    def _shape(self):
        if not _CH_RE.match(self.ch):
            raise ValueError(f"bad channel list {self.ch!r}")
        if not _COLOR_RE.match(self.color):
            raise ValueError(f"bad color {self.color!r}")
        n = len(self.ch)
        if not self.pts or len(self.pts) % n:
            raise ValueError("pts length must be a positive multiple of the channel count")
        return self

    @property
    def samples(self) -> int:
        return len(self.pts) // len(self.ch)


class InkFile(BaseModel):
    model_config = ConfigDict(extra="forbid")
    format: Literal["gamma-ink"]
    version: Literal[1]
    space: Space
    strokes: list[Stroke] = Field(default_factory=list, max_length=MAX_STROKES)

    @model_validator(mode="after")
    def _budget(self):
        if sum(s.samples for s in self.strokes) > MAX_SAMPLES:
            raise ValueError(f"more than {MAX_SAMPLES} samples")
        if len({s.id for s in self.strokes}) != len(self.strokes):
            raise ValueError("duplicate stroke ids")
        return self


class InkError(ValueError):
    pass


def parse_ink(data: bytes | str | dict) -> InkFile:
    """Validate bytes/JSON into an InkFile; InkError on anything off."""
    try:
        if isinstance(data, (bytes, bytearray)):
            if len(data) > MAX_BYTES:
                raise InkError(f"ink file over {MAX_BYTES // (1024 * 1024)} MB")
            data = json.loads(data.decode("utf-8"))
        elif isinstance(data, str):
            data = json.loads(data)
        return InkFile.model_validate(data)
    except (UnicodeDecodeError, json.JSONDecodeError, ValidationError) as e:
        raise InkError(str(e)) from e


def read_upload(uploads_dir, url: str) -> InkFile | None:
    """The parsed file a block's ``ink_url`` (``/api/uploads/<sha>.ink``)
    names, or None when the reference, the file or its contents are off;
    the exporters skip such a group."""
    from .markdown_export import UPLOAD_RE
    m = UPLOAD_RE.search(url or "")
    path = uploads_dir / m.group(1) if (m and uploads_dir) else None
    if not path or not path.is_file():
        return None
    try:
        return parse_ink(path.read_bytes())
    except (InkError, OSError):
        return None


# --- codec -------------------------------------------------------------------

def decode_stroke(stroke: Stroke) -> list[dict]:
    """A stroke's samples in points: ``[{x, y, p, t, a, z}]`` — ``p`` 0..1
    (0.5 when unrecorded), ``t`` ms since t0 or None, ``a``/``z`` degrees or
    None."""
    ch, pts, n = stroke.ch, stroke.pts, len(stroke.ch)
    out = []
    x = y = t = 0
    for i in range(0, len(pts), n):
        row = pts[i:i + n]
        rec = {"p": 0.5, "t": None, "a": None, "z": None}
        for c, v in zip(ch, row):
            if c == "x":
                x += v
                rec["x"] = x / COORD_UNIT
            elif c == "y":
                y += v
                rec["y"] = y / COORD_UNIT
            elif c == "p":
                rec["p"] = max(0.0, min(1.0, v / PRESSURE_UNIT))
            elif c == "t":
                t += v
                rec["t"] = t
            elif c == "a":
                rec["a"] = v
            elif c == "z":
                rec["z"] = v
        out.append(rec)
    return out


def encode_points(samples: list[dict], ch: str = "xy") -> list[int]:
    """The inverse of :func:`decode_stroke` (the importer and tests use it):
    ``samples`` carry ``x``/``y`` in points and the optional channels."""
    out = []
    px = py = pt = 0
    for s in samples:
        for c in ch:
            if c == "x":
                v = round(s["x"] * COORD_UNIT)
                out.append(v - px)
                px = v
            elif c == "y":
                v = round(s["y"] * COORD_UNIT)
                out.append(v - py)
                py = v
            elif c == "p":
                out.append(round(max(0.0, min(1.0, s.get("p", 0.5))) * PRESSURE_UNIT))
            elif c == "t":
                v = int(s.get("t") or 0)
                out.append(v - pt)
                pt = v
            elif c == "a":
                out.append(int(s.get("a") or 0))
            elif c == "z":
                out.append(int(s.get("z") or 0))
    return out


def stroke_width(stroke: Stroke, p: float) -> float:
    """Drawn diameter at pressure ``p``: pens thin with pressure, highlighters
    (and mouse/finger strokes, which carry no real pressure) stay constant."""
    if stroke.tool != "pen" or not stroke.pen:
        return stroke.size
    return stroke.size * (1 + THINNING * (p - 0.5))


def stroke_polyline(stroke: Stroke) -> list[tuple[float, float, float]]:
    """``[(x, y, width)]`` per sample — what every renderer here draws from
    (Xournal++'s model: variable-width polylines)."""
    return [(s["x"], s["y"], stroke_width(stroke, s["p"])) for s in decode_stroke(stroke)]


def bounding_box(ink: InkFile) -> tuple[float, float, float, float] | None:
    """(x0, y0, x1, y1) in points around every stroke, width included;
    None for an empty file."""
    x0 = y0 = math.inf
    x1 = y1 = -math.inf
    for stroke in ink.strokes:
        for x, y, w in stroke_polyline(stroke):
            r = w / 2
            x0, y0 = min(x0, x - r), min(y0, y - r)
            x1, y1 = max(x1, x + r), max(y1, y + r)
    if x0 is math.inf:
        return None
    return (x0, y0, x1, y1)


def pdf_position(ink: InkFile) -> dict | None:
    """The group's bounding box in the highlight ``pdf_position`` shape, so
    jump-to-position, markers and export anchoring treat ink like any other
    region on the page."""
    box = bounding_box(ink)
    if not box or ink.space.kind != "pdf-page":
        return None
    page = ink.space.page
    rect = {"x1": round(box[0], 2), "y1": round(box[1], 2), "x2": round(box[2], 2), "y2": round(box[3], 2),
            "width": ink.space.width, "height": ink.space.height, "pageNumber": page}
    return {"pageNumber": page, "boundingRect": rect, "rects": [dict(rect)]}


def parse_color(value: str) -> tuple[float, float, float, float]:
    """``#rrggbb`` or ``rgb[a](…)`` → (r, g, b, a) in 0..1."""
    v = (value or "").strip()
    if v.startswith("#") and len(v) == 7:
        return tuple(int(v[i:i + 2], 16) / 255 for i in (1, 3, 5)) + (1.0,)
    m = re.match(r"rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*(?:,\s*([0-9.]+)\s*)?\)", v)
    if m:
        r, g, b = (min(int(c), 255) / 255 for c in m.groups()[:3])
        return (r, g, b, min(float(m.group(4)), 1.0) if m.group(4) else 1.0)
    return (0.12, 0.12, 0.12, 1.0)


# --- renderers ------------------------------------------------------------------

def _fmt(v: float) -> str:
    return f"{v:.2f}".rstrip("0").rstrip(".")


def to_svg(ink: InkFile, box=None, pad: float = 4.0) -> str:
    """An SVG of the strokes cropped to ``box`` (default: the bounding box).
    Pens draw one round-capped segment per sample pair at that pair's width;
    highlighters one constant-width path with their opacity, multiplied onto
    what lies beneath."""
    box = box or bounding_box(ink) or (0, 0, 1, 1)
    x0, y0, x1, y1 = box[0] - pad, box[1] - pad, box[2] + pad, box[3] + pad
    w, h = max(1.0, x1 - x0), max(1.0, y1 - y0)
    parts = [f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="{_fmt(x0)} {_fmt(y0)} {_fmt(w)} {_fmt(h)}" '
             f'width="{_fmt(w)}" height="{_fmt(h)}">']
    for stroke in ink.strokes:
        poly = stroke_polyline(stroke)
        r, g, b, a = parse_color(stroke.color)
        color = f"rgb({round(r * 255)},{round(g * 255)},{round(b * 255)})"
        opacity = min(a, stroke.opacity)
        if stroke.tool == "highlighter":
            d = "M" + " L".join(f"{_fmt(x)} {_fmt(y)}" for x, y, _ in poly)
            if len(poly) == 1:
                d += f" L{_fmt(poly[0][0] + 0.01)} {_fmt(poly[0][1])}"
            parts.append(f'<path d="{d}" fill="none" stroke="{color}" stroke-width="{_fmt(stroke.size)}" '
                         f'stroke-opacity="{_fmt(opacity)}" stroke-linecap="round" stroke-linejoin="round" '
                         'style="mix-blend-mode:multiply"/>')
            continue
        parts.append(f'<g fill="none" stroke="{color}" stroke-linecap="round" stroke-linejoin="round"'
                     + (f' stroke-opacity="{_fmt(opacity)}"' if opacity < 1 else "") + ">")
        if len(poly) == 1:
            x, y, wd = poly[0]
            parts.append(f'<path d="M{_fmt(x)} {_fmt(y)} L{_fmt(x + 0.01)} {_fmt(y)}" stroke-width="{_fmt(wd)}"/>')
        for (ax, ay, aw), (bx, by, bw) in zip(poly, poly[1:]):
            parts.append(f'<path d="M{_fmt(ax)} {_fmt(ay)} L{_fmt(bx)} {_fmt(by)}" '
                         f'stroke-width="{_fmt((aw + bw) / 2)}"/>')
        parts.append("</g>")
    parts.append("</svg>")
    return "".join(parts)


def pdf_path_ops(ink: InkFile, to_pdf, scale: float = 1.0) -> bytes:
    """PDF content-stream operators drawing the strokes: ``to_pdf(x, y)``
    maps a display-space point to user space (the notes-as-PDF writer
    draws in a frame of its own). Highlighter opacity is baked into the
    colour (no ExtGState needed) — over white that is the same shade."""
    from .pdf_typeset import num
    ops = [b"q 1 J 1 j"]
    for stroke in ink.strokes:
        poly = stroke_polyline(stroke)
        r, g, b, a = parse_color(stroke.color)
        a = min(a, stroke.opacity)
        if a < 1:
            r, g, b = (c * a + (1 - a) for c in (r, g, b))
        ops.append(b"%s %s %s RG" % (num(r), num(g), num(b)))
        if stroke.tool == "highlighter":
            pts = [to_pdf(x, y) for x, y, _ in poly]
            ops.append(b"%s w" % num(stroke.size * scale))
            ops.append(b"%s %s m " % (num(pts[0][0]), num(pts[0][1]))
                       + b" ".join(b"%s %s l" % (num(x), num(y)) for x, y in pts[1:] or [pts[0]]) + b" S")
            continue
        for (ax, ay, aw), (bx, by, bw) in zip(poly, poly[1:] or poly):
            (px, py), (qx, qy) = to_pdf(ax, ay), to_pdf(bx, by)
            ops.append(b"%s w %s %s m %s %s l S" % (num((aw + bw) / 2 * scale), num(px), num(py), num(qx), num(qy)))
    ops.append(b"Q")
    return b"\n".join(ops)


# --- PDF /Ink interchange --------------------------------------------------------

def ink_buckets(ink: InkFile) -> list[list[Stroke]]:
    """Strokes grouped by look (tool, colour, size, opacity), in first-seen
    order: one ``/Ink`` annotation per bucket, since the PDF annotation has
    one width and one colour."""
    buckets: dict[tuple, list[Stroke]] = {}
    for s in ink.strokes:
        buckets.setdefault((s.tool, s.color.lower(), round(s.size, 2), round(s.opacity, 3)), []).append(s)
    return list(buckets.values())


def from_pdf_ink(ink_list, width_pt: float, color: str, opacity: float, page: int,
                 page_w: float, page_h: float, private_json: str | None = None) -> dict:
    """An embedded ``/Ink`` annotation → an ink file dict. ``ink_list`` is the
    ``/InkList`` (paths of x y pairs in PDF user space, origin bottom-left);
    ``private_json`` is the ``/GammaInk`` string a Gamma export left behind,
    which wins when it parses (it carries pressure and time)."""
    if private_json:
        try:
            ink = parse_ink(private_json)
            if ink.space.kind == "pdf-page":
                ink.space.page = page
            return ink.model_dump(exclude_none=True)
        except InkError:
            pass
    strokes = []
    for n, path in enumerate(ink_list or []):
        nums = [float(v) for v in path]
        samples = [{"x": nums[i], "y": page_h - nums[i + 1]} for i in range(0, len(nums) - 1, 2)]
        if not samples:
            continue
        strokes.append({"id": f"i{n}", "tool": "pen", "pen": False, "color": color,
                        "size": max(0.2, min(100.0, width_pt or 1.0)), "opacity": max(0.05, min(1.0, opacity)),
                        "ch": "xy", "pts": encode_points(samples, "xy")})
    return {"format": FORMAT, "version": VERSION,
            "space": {"kind": "pdf-page", "page": page, "width": page_w, "height": page_h},
            "strokes": strokes}


def dumps(ink: InkFile | dict) -> bytes:
    """Canonical bytes for storage (sorted keys, no whitespace) — the same
    strokes always hash to the same upload name."""
    data = ink.model_dump(exclude_none=True) if isinstance(ink, InkFile) else ink
    return json.dumps(data, separators=(",", ":"), sort_keys=True).encode("utf-8")
